include/zephyr/kernel.h

```diff
@@ -3606,6 +3606,19 @@ int k_work_queue_drain(struct k_work_q *queue, bool plug);
  */
 int k_work_queue_unplug(struct k_work_q *queue);
 
+/** @brief Stop a work queue.
+ *
+ * This stops the work queue thread and releases its resources. This should be
+ * invoked when the work queue is no longer needed.
+ *
+ * @param queue pointer to the queue structure.
+ *
+ * @retval 0 if the work queue was stopped
+ * @retval -ENODEV if the work queue was not started
+ * @retval -EBUSY if the work queue is actively processing work items
+ */
+int k_work_queue_stop(struct k_work_q *queue);
```

Comment:
Note that this is a blocking call (because of k_thread_join()) and so per "Standard Zephyr API Idiom" should probably take a k_timeout to allow the app to control wakeup behavior.  In this particular case that's probably not a big deal because the only way this will get stuck is if a workq handler refuses to return, which is already a bug elsewhere.

Reply:
I'm worried about side effects if an user sets a very short timeout, e.g. `K_NO_WAIT` and/or thread priorities. 
Where we get a false negative. 

The case I'm worried about is: 
```
int k_work_queue_stop(struct k_work_q *queue, k_timeout_t delay)
{
...
	flag_set(&queue->flags, K_WORK_QUEUE_STOP_BIT);
	notify_queue_locked(queue);
	k_spin_unlock(&lock, key);
	if (k_thread_join(&queue->thread, timeout)) { <--- sleeping here
		key = k_spin_lock(&lock);
		flag_clear(&queue->flags, K_WORK_QUEUE_STOP_BIT);
		k_spin_unlock(&lock, key);
		return -ETIMEOUT;
	}

	return 0;

...

static void work_queue_main(void *workq_ptr, void *p2, void *p3)
{
....
		k_spinlock_key_t key = k_spin_lock(&lock); <-- past this point
....		
}

		
k_work_queue_stop(queue, K_NO_WAIT); // returns -ETIMEOUT
// Is the thread on it's way to die or is it alive? 
```
It may still be ok? 
Subsequent calls to the workqueue will be handled (-ENODEV) due to the thread resets the flags.
I may be overthinking this. What do you think?

Reply:
If `k_work_queue_stop()` is guaranteed to set the `STOP` bit, then the work queue is guaranteed to stop in finite time (absent app bugs that stall the handlers).  I think that's acceptable.  Apps that care about synchronization can always call it again and take care to avoid racing with their own handlers until they get a success.

Again, I don't think this rises to a -1 given the specifics; the timeout isn't likely to be very useful in practice.  Nonetheless timeout arguments for blocking calls are The Zephyr Way, so it should be pointed out.